s3/src/main/scala/blobstore/s3/S3Store.scala

```diff
@@ -16,201 +16,419 @@ Copyright 2018 LendUp Global, Inc.
+          override def onResponse(response: GetObjectResponse): Unit = ()
+          override def onStream(publisher: SdkPublisher[ByteBuffer]): Unit = {
+            cf.complete(publisher.toStream().flatMap(bb => Stream.chunk(Chunk.byteBuffer(bb))))
+            ()
+          }
+          override def exceptionOccurred(error: Throwable): Unit = {
+            cf.completeExceptionally(error)
+            ()
           }
         }
-    }
+      Stream.eval(S3Store.liftJavaFuture(F.delay(s3.getObject[Stream[F, Byte]](request, transformer)))).flatten
   }
 
-  override def get(path: Path, chunkSize: Int): Stream[F, Byte] = {
-    val is: F[InputStream] = blocker.delay(s3.getObject(path.root, path.key).getObjectContent)
-    fs2.io.readInputStream(is, chunkSize, closeAfterUse = true, blocker = blocker)
-  }
+  override def put(path: Path): Pipe[F, Byte, Unit] =
+    in =>
+      S3Store.bucketKeyMetaFromPath(path) match {
+        case None => Stream.raiseError(S3Store.missingRootError(s"Unable to write to '$path'"))
+        case Some((bucket, key, meta)) =>
+          path.size.fold(putUnknownSize(bucket, key, meta, in))(size => putKnownSize(bucket, key, meta, size, in))
+      }
 
-  override def put(path: Path): Pipe[F, Byte, Unit] = { in =>
-    val init: F[(PipedOutputStream, PipedInputStream)] = F.delay {
-      val os = new PipedOutputStream()
-      val is = new PipedInputStream(os)
-      (os, is)
-    }
+  override def move(src: Path, dst: Path): F[Unit] =
+    copy(src, dst) *> remove(src)
 
-    val consume: ((PipedOutputStream, PipedInputStream)) => Stream[F, Unit] = ios => {
-      val putToS3 = Stream.eval(blocker.delay {
-        val meta = new ObjectMetadata()
-        path.size.foreach(meta.setContentLength)
-        sseAlgorithm.foreach(meta.setSSEAlgorithm)
-        transferManager.upload(path.root, path.key, ios._2, meta).waitForCompletion()
-        objectAcl.foreach(acl => s3.setObjectAcl(path.root, path.key, acl))
-        ()
-      })
+  override def copy(src: Path, dst: Path): F[Unit] =
+    for {
+      (srcBucket, srcKey, _) <- S3Store
+        .bucketKeyMetaFromPath(src)
+        .fold[F[(String, String, Option[S3MetaInfo])]](S3Store.missingRootError(s"Wrong src '$src'").raiseError)(
+          _.pure[F]
+        )
+      (dstBucket, dstKey, dstMeta) <- S3Store
+        .bucketKeyMetaFromPath(dst)
+        .fold[F[(String, String, Option[S3MetaInfo])]](S3Store.missingRootError(s"Wrong dst '$dst'").raiseError)(
+          _.pure[F]
+        )
+      _ <- {
+        val request = dstMeta
+          .fold {
+            val builder = CopyObjectRequest.builder()
+            val withAcl = objectAcl.fold(builder)(builder.acl)
+            sseAlgorithm.fold(withAcl)(withAcl.serverSideEncryption)
+          }(S3MetaInfo.copyObjectRequest)
+          .copySource(srcBucket ++ "/" ++ srcKey)
+          .destinationBucket(dstBucket)
+          .destinationKey(dstKey)
```

Comment:
Missing SSE configuration

Reply:
For paths which are not `S3Path` acl and sseAlgorithm are set above in the empty case of fold.
For paths which are `S3Path` they are set based on `S3MetaInfo` in those request builders `S3MetaInfo.copyObjectRequest` etc.

It might be a good idea to set default acl and sseAlgorithm, if it they are not set in `S3MetaInfo`, but values in `S3MetaInfo` should take precedence.